interpreter/code_interpreters/container_utils/container_utils.py

```diff
@@ -0,0 +1,313 @@
+# Standard library imports
+import atexit
+import hashlib
+import json
+import os
+import re
+import select
+import shutil
+import struct
+import subprocess
+import threading
+import time
+
+
+# Third-party imports
+import docker
+from docker import DockerClient
+from docker.errors import DockerException
+from rich import print as Print
+
+
+def get_files_hash(*file_paths):
+    """Return the SHA256 hash of multiple files."""
+    hasher = hashlib.sha256()
+    for file_path in file_paths:
+        with open(file_path, "rb") as f:
+            while chunk := f.read(4096):
+                hasher.update(chunk)
+    return hasher.hexdigest()
+
+
+def build_docker_images(
+    dockerfile_dir = os.path.join(os.path.abspath(os.path.dirname(os.path.dirname(__file__))), "dockerfiles")
+,
+):
+    """
+    Builds a Docker image for the Open Interpreter runtime container if needed.
+
+    Args:
+        dockerfile_dir (str): The directory containing the Dockerfile and requirements.txt files.
+
+    Returns:
+        None
+    """
+    try:
+        client = DockerClient.from_env()
+    except DockerException:
+        Print("ERROR: Could not connect to Docker daemon. Is Docker Engine installed and running?")
+        Print(
+            "\nFor information on Docker installation, visit: https://docs.docker.com/engine/install/"
+        )
+        return
+
+    image_name = "openinterpreter-runtime-container"
+    hash_file_path = os.path.join(dockerfile_dir, "hash.json")
+
+    dockerfile_name = "Dockerfile"
+    requirements_name = "requirements.txt"
+    dockerfile_path = os.path.join(dockerfile_dir, dockerfile_name)
+    requirements_path = os.path.join(dockerfile_dir, requirements_name)
+
+    if not os.path.exists(dockerfile_path) or not os.path.exists(requirements_path):
+        Print("ERROR: Dockerfile or requirements.txt not found. Did you delete or rename them?")
+        raise RuntimeError(
+            "No container Dockerfiles or requirements.txt found. Make sure they are in the dockerfiles/ subdir of the module."
+        )
+
+    current_hash = get_files_hash(dockerfile_path, requirements_path)
+
+    stored_hashes = {}
+    if os.path.exists(hash_file_path):
+        with open(hash_file_path, "rb") as f:
+            stored_hashes = json.load(f)
+
+    original_hash = stored_hashes.get("original_hash")
+    previous_hash = stored_hashes.get("last_hash")
+
+    if current_hash == original_hash:
+        images = client.images.list(name=image_name, all=True)
+        if not images:
+            Print("Downloading default image from Docker Hub, please wait...")
```

Comment:
We might want to put this behind `debug_mode` as it seems to pop up more often than I would have expected.

Reply:
huh. it wasnt doing that before. lemme look into it. I think I know whats causing it.

Reply:
this has been resolved. docker was tagging the image with the repository name when pulling it from dockerhub, but I had it configured to look for it locally. we now re-tag the image using ```docker tag [repo image name] [new local image name]```

I will push this in my next commit.